src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_ioctl.c

```diff
@@ -349,6 +329,39 @@ xocl_read_axlf_helper(struct xocl_drm *drm_p, struct drm_xocl_axlf *axlf_ptr)
 	if (xclbin_downloaded(xdev, &bin_obj.m_header.uuid))
 		goto done;
 
+	/*
```

Comment:
i don't see anything wrong to separate xclbin download and xclbin lock/unlock operation. but i agree that the coupling here is not ideal.

Reply:
I agree. Separate download and lock is okay. I will change the comment.
If anyone try to load another xclbin before KDS unlock prev xclbin, ICAP would failed here. If it does need to check KDS status in here, we could improve this in the future.